File: src/main/java/com/hileco/cortex/vm/layer/LayeredBytes.java
package com.hileco.cortex.vm.layer;

import java.util.Arrays;

public class LayeredBytes {

    public static final int DEFAULT_TOTAL_SIZE = 8192;
    private byte[] bytes;

    public LayeredBytes() {
        this.bytes = new byte[DEFAULT_TOTAL_SIZE];
    }

    public byte[] read(int offset, int length) {
        return Arrays.copyOfRange(this.bytes, offset, offset + length);
    }

    public void clear() {
        this.bytes = new byte[DEFAULT_TOTAL_SIZE];
    }

    public void write(int offset, byte[] bytesToWrite) {
        this.write(offset, bytesToWrite, bytesToWrite.length);
    }

    public void write(int offset, byte[] bytesToWrite, int writeLength) {
        System.arraycopy(bytesToWrite, 0, this.bytes, offset, writeLength);
    }

    public LayeredBytes copy() {
        throw new UnsupportedOperationException("Copying LayeredBytes is currently not supported.");
    }

    @Override
    public String toString() {
        return String.format("LayeredBytes{size %d}", this.bytes.length);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || this.getClass() != o.getClass()) return false;
        var that = (LayeredBytes) o;
        return Arrays.equals(this.bytes, that.bytes);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(this.bytes);
    }
}

File: src/main/java/com/hileco/cortex/vm/layer/LayeredBytes.kt
package com.hileco.cortex.vm.layer

import java.util.*

class LayeredBytes(val size: Int = DEFAULT_TOTAL_SIZE) : DelegateLayered<LayeredBytes>() {
    private val lazyBytes = lazy { ByteArray(size) }
    val bytes: ByteArray by lazyBytes

    @Synchronized
    fun read(offset: Int, length: Int): ByteArray {
        return Arrays.copyOfRange(bytes, offset, offset + length)
    }

    @Synchronized
    fun clear() {
        System.arraycopy(ByteArray(size), 0, bytes, 0, size)
    }

    @Synchronized
    fun write(offset: Int, bytesToWrite: ByteArray, writeLength: Int = bytesToWrite.size) {
        System.arraycopy(bytesToWrite, 0, bytes, offset, writeLength)
    }

    @Synchronized
    override fun recreateParent(): LayeredBytes {
        return LayeredBytes(size)
    }

    @Synchronized
    override fun branchDelegates(): LayeredBytes {
        val clone = LayeredBytes(size)
        clone.write(0, bytes)
        return clone
    }

    @Synchronized
    override fun disposeDelegates() {
    }

    override fun toString(): String {
        return "LayeredBytes(size=$size)"
    }

    override fun equals(other: Any?): Boolean {
        return other != null && other is LayeredBytes && Arrays.equals(bytes, other.bytes)
    }

    override fun hashCode(): Int {
        return size.hashCode()
    }

    companion object {
        const val DEFAULT_TOTAL_SIZE = 262144
    }
}
